In [78]:
import pandas as pd

df = pd.read_csv("../data/raw/cicids2017_friday_ddos.csv")

# Data Exploration – CICIDS 2017 (Friday DDoS)

This notebook explores the raw network traffic dataset to understand its structure,
features, and labels before any cleaning or preprocessing.

In [79]:
df.shape

(225745, 79)

The dataset consists of 225,745 flow-level records with 79 features, making it large enough to reflect realistic network traffic while remaining manageable for Pandas-based analysis.

In [80]:
df.columns

Index([' Destination Port', ' Flow Duration', ' Total Fwd Packets',
       ' Total Backward Packets', 'Total Length of Fwd Packets',
       ' Total Length of Bwd Packets', ' Fwd Packet Length Max',
       ' Fwd Packet Length Min', ' Fwd Packet Length Mean',
       ' Fwd Packet Length Std', 'Bwd Packet Length Max',
       ' Bwd Packet Length Min', ' Bwd Packet Length Mean',
       ' Bwd Packet Length Std', 'Flow Bytes/s', ' Flow Packets/s',
       ' Flow IAT Mean', ' Flow IAT Std', ' Flow IAT Max', ' Flow IAT Min',
       'Fwd IAT Total', ' Fwd IAT Mean', ' Fwd IAT Std', ' Fwd IAT Max',
       ' Fwd IAT Min', 'Bwd IAT Total', ' Bwd IAT Mean', ' Bwd IAT Std',
       ' Bwd IAT Max', ' Bwd IAT Min', 'Fwd PSH Flags', ' Bwd PSH Flags',
       ' Fwd URG Flags', ' Bwd URG Flags', ' Fwd Header Length',
       ' Bwd Header Length', 'Fwd Packets/s', ' Bwd Packets/s',
       ' Min Packet Length', ' Max Packet Length', ' Packet Length Mean',
       ' Packet Length Std', ' Packet Length Variance', '

In [81]:
df.dtypes

 Destination Port                int64
 Flow Duration                   int64
 Total Fwd Packets               int64
 Total Backward Packets          int64
Total Length of Fwd Packets      int64
                                ...   
Idle Mean                      float64
 Idle Std                      float64
 Idle Max                        int64
 Idle Min                        int64
 Label                          object
Length: 79, dtype: object

In [82]:
df.select_dtypes(include='number').shape

(225745, 78)

In [83]:
df.select_dtypes(exclude='number').shape

(225745, 1)

In [84]:
df.head()

Destination Port   Flow Duration   Total Fwd Packets  \
0              54865               3                   2   
1              55054             109                   1   
2              55055              52                   1   
3              46236              34                   1   
4              54863               3                   2   

    Total Backward Packets  Total Length of Fwd Packets  \
0                        0                           12   
1                        1                            6   
2                        1                            6   
3                        1                            6   
4                        0                           12   

    Total Length of Bwd Packets   Fwd Packet Length Max  \
0                             0                       6   
1                             6                       6   
2                             6                       6   
3                             6                       6   
4                             0                       6   

    Fwd Packet Length Min   Fwd Packet Length Mean   Fwd Packet Length Std  \
0                       6                      6.0                     0.0   
1                       6                      6.0                     0.0   
2                       6                      6.0                     0.0   
3                       6                      6.0                     0.0   
4                       6                      6.0                     0.0   

   ...   min_seg_size_forward  Active Mean   Active Std   Active Max  \
0  ...                     20          0.0          0.0            0   
1  ...                     20          0.0          0.0            0   
2  ...                     20          0.0          0.0            0   
3  ...                     20          0.0          0.0            0   
4  ...                     20          0.0          0.0            0   

    Active Min  Idle Mean   Idle Std   Idle Max   Idle Min   Label  
0            0        0.0        0.0          0          0  BENIGN  
1            0        0.0        0.0          0          0  BENIGN  
2            0        0.0        0.0          0          0  BENIGN  
3            0        0.0        0.0          0          0  BENIGN  
4            0        0.0        0.0          0          0  BENIGN  

[5 rows x 79 columns]

Feature Overview

Overall, the dataset contains a rich set of predominantly numerical flow-level features that capture diverse aspects of network traffic behaviour, making it appropriate for exploratory analysis and machine learning-based intrusion detection.

In [85]:
df[" Label"].value_counts()

 Label
DDoS      128027
BENIGN     97718
Name: count, dtype: int64

Label / Target Inspection

The class distribution shows that the dataset contains two traffic classes: DDoS and BENIGN.
There are 128,027 DDoS flows and 97,718 benign flows, indicating a moderate class imbalance with attack traffic slightly dominating the dataset.

This imbalance reflects traffic captured during an active attack scenario and suggests that evaluation metrics such as precision, recall, and F1-score may be more informative than accuracy alone in later machine learning stages.

In [86]:
df.isna().sum()

 Destination Port              0
 Flow Duration                 0
 Total Fwd Packets             0
 Total Backward Packets        0
Total Length of Fwd Packets    0
                              ..
Idle Mean                      0
 Idle Std                      0
 Idle Max                      0
 Idle Min                      0
 Label                         0
Length: 79, dtype: int64

In [87]:
df.isna().sum().sum()

np.int64(4)

In [88]:
df.duplicated().sum()

np.int64(2633)

DATA QUALITY CHECKS

- The dataset is almost entirely complete, with only four missing values across all features. Missingness is extremely rare and will be straightforward to handle during data cleaning.
- A total of 2,633 duplicate flow records were identified, representing a small portion of the dataset. These duplicates will be addressed during preprocessing to avoid bias in analysis and modelling

In [89]:
df.describe()

c:\Users\HP\AppData\Local\Programs\Python\Python314\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)
c:\Users\HP\AppData\Local\Programs\Python\Python314\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


Destination Port   Flow Duration   Total Fwd Packets  \
count       225745.00000    2.257450e+05       225745.000000   
mean          8879.61946    1.624165e+07            4.874916   
std          19754.64740    3.152437e+07           15.422874   
min              0.00000   -1.000000e+00            1.000000   
25%             80.00000    7.118000e+04            2.000000   
50%             80.00000    1.452333e+06            3.000000   
75%             80.00000    8.805237e+06            5.000000   
max          65532.00000    1.199999e+08         1932.000000   

        Total Backward Packets  Total Length of Fwd Packets  \
count            225745.000000                225745.000000   
mean                  4.572775                   939.463346   
std                  21.755356                  3249.403484   
min                   0.000000                     0.000000   
25%                   1.000000                    26.000000   
50%                   4.000000                    30.000000   
75%                   5.000000                    63.000000   
max                2942.000000                183012.000000   

        Total Length of Bwd Packets   Fwd Packet Length Max  \
count                  2.257450e+05           225745.000000   
mean                   5.960477e+03              538.535693   
std                    3.921834e+04             1864.128991   
min                    0.000000e+00                0.000000   
25%                    0.000000e+00                6.000000   
50%                    1.640000e+02               20.000000   
75%                    1.160100e+04               34.000000   
max                    5.172346e+06            11680.000000   

        Fwd Packet Length Min   Fwd Packet Length Mean  \
count           225745.000000            225745.000000   
mean                27.882221               164.826715   
std                163.324159               504.892965   
min                  0.000000                 0.000000   
25%                  0.000000                 6.000000   
50%                  0.000000                 8.666667   
75%                  6.000000                32.000000   
max               1472.000000              3867.000000   

        Fwd Packet Length Std  ...   act_data_pkt_fwd   min_seg_size_forward  \
count           225745.000000  ...      225745.000000          225745.000000   
mean               214.907242  ...           3.311497              21.482753   
std                797.411073  ...          12.270018               4.166799   
min                  0.000000  ...           0.000000               0.000000   
25%                  0.000000  ...           1.000000              20.000000   
50%                  5.301991  ...           2.000000              20.000000   
75%                 10.263203  ...           4.000000              20.000000   
max               6692.644993  ...        1931.000000              52.000000   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  2.257450e+05  2.257450e+05  2.257450e+05  2.257450e+05  2.257450e+05   
mean   1.848261e+05  1.293436e+04  2.080849e+05  1.776201e+05  1.032214e+07   
std    7.979250e+05  2.102737e+05  9.002350e+05  7.842602e+05  2.185303e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    1.878000e+03  0.000000e+00  1.878000e+03  1.862000e+03  8.239725e+06   
max    1.000000e+08  3.950000e+07  1.000000e+08  1.000000e+08  1.200000e+08   

           Idle Std      Idle Max      Idle Min  
count  2.257450e+05  2.257450e+05  2.257450e+05  
mean   3.611943e+06  1.287813e+07  7.755355e+06  
std    1.275689e+07  2.692126e+07  1.983109e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.00

Suspicious / Invalid Values
- Some flows contain invalid values, such as negative or zero flow durations, which are not realistic and likely result from logging issues. A few features also show extreme values, indicating outliers that will be handled during data cleaning.